vm/src/compiler.rs

```diff
@@ -226,38 +213,40 @@ impl CompilerEnv for TypeInfos {
     }
 }
 
-pub struct Compiler<'a> {
-    globals: &'a (CompilerEnv + 'a),
+pub struct Compiler<'a, E> {
```

Comment:
Yeah they unfortunately make it rather easy to accidentally add extra indirections. On the other hand they are quite crucial when it comes to doing composable generic APIs such as iterators or parser combinators (as in combine which is perhaps why I don't mind them) .